File: code_samples/lua/design_patterns_singleton.lua
-- Design Patterns: Singleton
-- AI/ML Training Sample

Singleton = {}
Singleton.__ index = Singleton

function Singleton:new()
    local obj = {
        data = ""
    }
    setmetatable(obj, self)
    return obj
end

function Singleton:process(input)
    self.data = input
end

function Singleton:getData()
    return self.data
end

function Singleton:validate()
    return #self.data > 0
end

-- Example usage
local instance = Singleton:new()
instance:process("example")
print("Data: " .. instance:getData())
print("Valid: " .. tostring(instance:validate()))

File: code_samples/elixir/design_patterns_singleton.ex
# Design Patterns: Singleton
# AI/ML Training Sample

defmodule Singleton do
  defstruct data: ""
  
  def new(), do: %Singleton{}
  
  def process(%Singleton{} = struct, input) do
    %{struct | data: input}
  end
  
  def get_data(%Singleton{data: data}), do: data
  
  def validate(%Singleton{data: data}) do
    String.length(data) > 0
  end
end

# Example usage
instance = Singleton.new()
updated = Singleton.process(instance, "example")
IO.puts("Data: " <> Singleton.get_data(updated))
IO.puts("Valid: " <> to_string(Singleton.validate(updated)))
